src/libstore/build/local-derivation-goal.cc

```diff
@@ -2474,7 +2474,13 @@ void LocalDerivationGoal::registerOutputs()
         infos.emplace(outputName, std::move(newInfo));
     }
 
-    if (buildMode == bmCheck) return;
+    if (buildMode == bmCheck) {
+        // In case of FOD mismatches on `--check` an error must be thrown as this is also
+        // a source for non-determinism.
+        if (delayedException)
+            std::rethrow_exception(delayedException);
```

Comment:
It’d be nice in that case to register the invalid path before throwing (like we do when not `--check`-ing). But I’m not sure whether that can be done without complicating the code a lot, so I’ll leave that up to you

Reply:
Which code-path do you refer to? I guess it makes sense since otherwise this path could be re-used in a later build, or am I misunderstanding this? :)

Reply:
What I say is that when `--check` is used and there’s a hash mismatch, then the mismatched path isn’t registered at all (and as you say, it would make sense to do so). But it’s not really critical given that it already wasn’t the case